File: REST_DOTNET/04. CLICON/Controlador/ControladorConversiones.cs
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;

namespace ClienteConsola.Controlador
{
    public class ControladorConversiones
    {
        private readonly HttpClient _httpClient;
        private string? _token;

        public ControladorConversiones()
        {
            _httpClient = new HttpClient
            {
                BaseAddress = new Uri("https://server_rest.dr00p3r.top/")
            };
        }

        public bool iniciarSesion(string usuario, string clave)
        {
            try
            {
                var solicitud = new { Usuario = usuario, Clave = clave };
                var json = JsonSerializer.Serialize(solicitud);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var response = _httpClient.PostAsync("api/auth/login", content).GetAwaiter().GetResult();
                var result = response.Content.ReadFromJsonAsync<LoginResponse>().GetAwaiter().GetResult();

                if (result != null && result.Exito)
                {
                    _token = result.TokenGenerado;
                    return true;
                }
                return false;
            }
            catch
            {
                return false;
            }
        }

        public double convertirLongitud(double valor, string origen, string destino)
        {
            return EjecutarConversion("api/conversion/longitud", valor, origen, destino);
        }

        public double convertirTemperatura(double valor, string origen, string destino)
        {
            return EjecutarConversion("api/conversion/temperatura", valor, origen, destino);
        }

        public double convertirMasa(double valor, string origen, string destino)
        {
            return EjecutarConversion("api/conversion/masa", valor, origen, destino);
        }

        public bool validarCampoNumerico(string valorStr)
        {
            if (string.IsNullOrEmpty(_token))
                throw new InvalidOperationException("No hay sesion activa.");

            var solicitud = new { ValorStr = valorStr };
            var json = JsonSerializer.Serialize(solicitud);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var request = new HttpRequestMessage(HttpMethod.Post, "api/conversion/validar-numerico")
            {
                Content = content
            };
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);

            var response = _httpClient.SendAsync(request).GetAwaiter().GetResult();
            var result = response.Content.ReadFromJsonAsync<ValidarNumericoResponse>().GetAwaiter().GetResult();

            if (result == null || !result.Exito)
                throw new Exception(result?.MensajeError ?? "Error desconocido");

            return result.EsValido;
        }

        private double EjecutarConversion(string endpoint, double valor, string origen, string destino)
        {
            if (string.IsNullOrEmpty(_token))
                throw new InvalidOperationException("No hay sesion activa.");

            var solicitud = new { Valor = valor, UnidadOrigen = origen, UnidadDestino = destino };
            var json = JsonSerializer.Serialize(solicitud);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var request = new HttpRequestMessage(HttpMethod.Post, endpoint)
            {
                Content = content
            };
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);

            var response = _httpClient.SendAsync(request).GetAwaiter().GetResult();
            var result = response.Content.ReadFromJsonAsync<ConversionResponse>().GetAwaiter().GetResult();

            if (result == null || !result.Exito)
                throw new Exception(result?.MensajeError ?? "Error desconocido");

            return result.ValorConvertido;
        }

        private class LoginResponse
        {
            public bool Exito { get; set; }
            public string TokenGenerado { get; set; } = string.Empty;
            public string MensajeError { get; set; } = string.Empty;
        }

        private class ConversionResponse
        {
            public bool Exito { get; set; }
            public double ValorConvertido { get; set; }
            public string MensajeError { get; set; } = string.Empty;
        }

        private class ValidarNumericoResponse
        {
            public bool Exito { get; set; }
            public bool EsValido { get; set; }
            public string MensajeError { get; set; } = string.Empty;
        }
    }
}

File: REST_JAVA/04. CLIESC/ClienteEscritorio/src/main/java/ec/edu/monster/clienteescritorio/controlador/ControladorConversiones.java
package ec.edu.monster.clienteescritorio.controlador;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 *
 * @author joanc
 */
public class ControladorConversiones {
    private String token;
    // URL directa sin el /api/
    private static final String BASE_URL = "https://javarest.dr00p3r.top/WS_CONV_UNI_RESTFULL_JAVA";
    private final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();

    public boolean iniciarSesion(String usuario, String clave) {
        // Armamos el JSON manual para no meter librerias extra (Java Purito)
        String json = String.format("{\"from\":\"%s\", \"to\":\"%s\"}", usuario, clave);

        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(BASE_URL + "/auth/login"))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(json))
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() == 200) {
                // Parseo manual rapido para sacar el token del JSON
                this.token = response.body().split("\"token\":\"")[1].split("\"")[0];
                return true;
            } else {
                return false;
            }
        } catch (Exception e) {
            System.err.println("Error en comunicacion REST: " + e.getMessage());
            return false;
        }
    }

    // Metodo privado para reutilizar la logica de peticiones POST
    private double ejecutarPeticion(String path, double valor, String origen, String destino) throws Exception {
        String json = String.format("{\"value\":%s, \"from\":\"%s\", \"to\":\"%s\", \"token\":\"%s\"}",
                valor, origen, destino, this.token);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + "/conversor/" + path))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        String body = response.body();

        if (response.statusCode() == 200) {
            // Extraemos el valor del campo "result"
            String res = body.split("\"result\":")[1].split(",")[0];
            return Double.parseDouble(res);
        } else {
            // Si el servidor mando un error (ej. unidades invalidas), sacamos el mensaje
            String errorMsg = body.split("\"message\":\"")[1].split("\"")[0];
            throw new Exception(errorMsg);
        }
    }

    public double convertirLongitud(double valor, String origen, String destino) throws Exception {
        return ejecutarPeticion("longitud", valor, origen, destino);
    }

    public double convertirTemperatura(double valor, String origen, String destino) throws Exception {
        return ejecutarPeticion("temperatura", valor, origen, destino);
    }

    public double convertirMasa(double valor, String origen, String destino) throws Exception {
        return ejecutarPeticion("masa", valor, origen, destino);
    }
}